## getting RGBA colors

RGBA - each one is an integer from 0 (no color) to 255 (max color), each one represented using 8-bit unsigned integers. A is alpha value, representing transparency.
We pass a tuple with four int values to Pillow. For example Blue is represented by `(0, 0, 255, 255)`

The **ImageColor** module contains color tables and converters from CSS3-style color specifiers to RGB tuples. You can use hexadecimal color specifiers to get the RGB values from the ImageColor module. Pillow also supports Hue-Saturation-Lightness (HSL) and Hue-Saturation-Value (HSV) functions.

In [3]:
from PIL import ImageColor
ImageColor.getcolor('red', 'RGBA')

(255, 0, 0, 255)

In [9]:
ImageColor.getrgb('hsv(0, 100%, 100%)')

(255, 0, 0)

get a full listing of the named colors that Pillow supports by accessing the colormap attribute of the ImageColor module. it will return colors as a Python dictionary. The keys of the dictionary are the HTML color names while the values are the corresponding hexadecimal representations of the colors. Colors are all in RGB, not in RGBA

In [10]:
ImageColor.colormap

{'aliceblue': '#f0f8ff',
 'antiquewhite': '#faebd7',
 'aqua': '#00ffff',
 'aquamarine': '#7fffd4',
 'azure': '#f0ffff',
 'beige': '#f5f5dc',
 'bisque': '#ffe4c4',
 'black': '#000000',
 'blanchedalmond': '#ffebcd',
 'blue': '#0000ff',
 'blueviolet': '#8a2be2',
 'brown': '#a52a2a',
 'burlywood': '#deb887',
 'cadetblue': '#5f9ea0',
 'chartreuse': '#7fff00',
 'chocolate': '#d2691e',
 'coral': '#ff7f50',
 'cornflowerblue': '#6495ed',
 'cornsilk': '#fff8dc',
 'crimson': '#dc143c',
 'cyan': '#00ffff',
 'darkblue': '#00008b',
 'darkcyan': '#008b8b',
 'darkgoldenrod': '#b8860b',
 'darkgray': '#a9a9a9',
 'darkgrey': '#a9a9a9',
 'darkgreen': '#006400',
 'darkkhaki': '#bdb76b',
 'darkmagenta': '#8b008b',
 'darkolivegreen': '#556b2f',
 'darkorange': '#ff8c00',
 'darkorchid': '#9932cc',
 'darkred': '#8b0000',
 'darksalmon': '#e9967a',
 'darkseagreen': '#8fbc8f',
 'darkslateblue': '#483d8b',
 'darkslategray': '#2f4f4f',
 'darkslategrey': '#2f4f4f',
 'darkturquoise': '#00ced1',
 'darkviolet': '#9400d3

In [11]:
len(ImageColor.colormap)

148

Use colormap in a list comprehension to get the RBGA value for each of the colors.

In [14]:
rgba_colors = { name: ImageColor.getcolor(name, 'RGBA') for name in ImageColor.colormap }
print(rgba_colors)

{'aliceblue': (240, 248, 255, 255), 'antiquewhite': (250, 235, 215, 255), 'aqua': (0, 255, 255, 255), 'aquamarine': (127, 255, 212, 255), 'azure': (240, 255, 255, 255), 'beige': (245, 245, 220, 255), 'bisque': (255, 228, 196, 255), 'black': (0, 0, 0, 255), 'blanchedalmond': (255, 235, 205, 255), 'blue': (0, 0, 255, 255), 'blueviolet': (138, 43, 226, 255), 'brown': (165, 42, 42, 255), 'burlywood': (222, 184, 135, 255), 'cadetblue': (95, 158, 160, 255), 'chartreuse': (127, 255, 0, 255), 'chocolate': (210, 105, 30, 255), 'coral': (255, 127, 80, 255), 'cornflowerblue': (100, 149, 237, 255), 'cornsilk': (255, 248, 220, 255), 'crimson': (220, 20, 60, 255), 'cyan': (0, 255, 255, 255), 'darkblue': (0, 0, 139, 255), 'darkcyan': (0, 139, 139, 255), 'darkgoldenrod': (184, 134, 11, 255), 'darkgray': (169, 169, 169, 255), 'darkgrey': (169, 169, 169, 255), 'darkgreen': (0, 100, 0, 255), 'darkkhaki': (189, 183, 107, 255), 'darkmagenta': (139, 0, 139, 255), 'darkolivegreen': (85, 107, 47, 255), 'darko

## getting colors from images

We can get a color from an image by using `getcolors()` the default limit of colors is 256, if the number of colors exceeds it, it returns `None`, we can increase `maxcolors`. It returns an unsorted list of (count, pixel) values.

In [37]:
from PIL import Image

image = Image.open('cat.jpg')
# image.show("Cat")
colors = image.getcolors(maxcolors=256000)
print(colors)

# with Image.open('cat.jpg') as image:
#   colors = image.getcolors()
#   print(colors)



[(1, (255, 255, 255)), (663, (254, 255, 255)), (48, (252, 255, 255)), (89, (251, 255, 255)), (176, (250, 255, 255)), (176, (249, 255, 255)), (92, (248, 255, 255)), (420, (247, 255, 255)), (70, (246, 255, 255)), (50, (245, 255, 255)), (13, (244, 255, 251)), (11, (243, 255, 255)), (7, (242, 255, 255)), (10, (241, 255, 255)), (1, (240, 255, 255)), (2, (239, 255, 255)), (20, (3, 0, 0)), (113, (8, 0, 0)), (18, (244, 255, 255)), (4, (94, 1, 32)), (211, (4, 0, 0)), (18, (76, 1, 32)), (4, (12, 0, 0)), (41, (5, 0, 0)), (2, (11, 1, 0)), (243, (2, 1, 0)), (116, (6, 0, 0)), (51, (86, 1, 32)), (1, (141, 149, 136)), (16, (70, 1, 32)), (8, (97, 7, 32)), (4, (20, 0, 0)), (1, (21, 0, 0)), (1, (29, 0, 0)), (1, (37, 0, 0)), (1, (7, 2, 0)), (2, (100, 6, 32)), (2, (101, 5, 32)), (2, (69, 7, 32)), (25, (98, 5, 32)), (5, (73, 0, 32)), (21, (74, 4, 32)), (3, (74, 0, 32)), (8, (3, 4, 0)), (6, (79, 0, 32)), (13, (87, 4, 32)), (19, (79, 5, 32)), (12, (90, 0, 32)), (7, (133, 137, 146)), (5, (89, 0, 32)), (1, (88,

### changing pixel colors

let's create a blank image first

In [34]:
stripe_image = Image.new("RGBA", (200, 200))

red = ImageColor.getcolor('red', 'RGBA')
green = ImageColor.getcolor('green', 'RGBA')

color = red

count = 0
for y in range(200):
    for x in range(200):
        if count == 5:
            # swap colors
            color = red if red != color else green
            count = 0
        stripe_image.putpixel((x, y), color)
        count += 1

stripe_image.save('rg_stripes.png')
stripe_image.show("rg_stripes")

In [36]:
print(red)

(255, 0, 0, 255)


Pillow is not made for fast pixel-by-pixel editing. If you want to try to do this on a high-resolution image, you should use NumPy instead.

### color transforms with `convert()`

Parameters:
- *mode*
  - "1" - 1-bit pixels
  - "L" - 8-bit, grayscale
  - "P" - 8-bit, mapped to any other mode using a color palette
  - "RGB" - 3x8-bit, true color
  - "RGBA" - 4x8-bit, true color + alpha
  - "CMYK" - 4x8-bit, color separation
  - "YCbCr" - 3x8-bit, color video format
  - "LAB" - 3x8-bit, Lab color space
  - "HSV" - 3x8 bit, hsv color space
  - "I" - 32-bit signed integer pixels
  - "F" - 32-bit floating point pixels
- *matrix* an optional conversion matrix
- *dither* dithering method when converting to "P" or "1", not used when matrix is supplied
  - `Dither.NONE`
  - `Dither.FLOYDSTEINBERG`default
- *palette* used when converting from "RGB" to "P"
  - `Palette.WEB`
  - `Palette.ADAPTIVE`
- *colors* num of colors used for `Palette.ADAPTIVE`, defaults to 256

When translating a color image to grayscale (mode “L”), the library uses the ITU-R 601-2 luma transform:
`L = R * 299/1000 + G * 587/1000 + B * 114/1000`

The default method of converting a grayscale (“L”) or “RGB” image into a bilevel (mode “1”) image uses **Floyd-Steinberg dither** to approximate the original image luminosity levels. If dither is None, all values larger than 127 are set to 255 (white), all other values to 0 (black). To use other thresholds, use the point() method.

In [63]:
# convert to bitmap with Floyd-Steinberg dithering
bw_image = image.convert("1", dither=3)
# bw_image.show()

# convert from RGB to CIE XYZ using a matrix
rgb2xyz = (
    0.412453, 0.357580, 0.180423, 0,
    0.212671, 0.715160, 0.072169, 0,
    0.019334, 0.119193, 0.950227, 0)
out = image.convert("RGB", rgb2xyz)
# out.show()

# convert to 16 color palette and save (show() doesn't work)
plt_im = image.convert("P", palette=Image.ADAPTIVE, colors=16)
plt_im.save("16color_cat.png")

`Image.putpalette(data, rawmode='RGB')`
PARAMETERS:
data – A palette sequence (either a list or a string).
rawmode – The raw mode of the palette. Either “RGB”, “RGBA”, or a mode that can be transformed to “RGB” or “RGBA”

Attaches a palette to this image. The image must be a “P”, “PA”, “L” or “LA” image.
The palette sequence must contain at most 256 colors, made up of one integer value for each channel in the raw mode. For example, if the raw mode is “RGB”, then it can contain at most 768 values, made up of red, green and blue values for the corresponding pixel index in the 256 colors.

Alternatively, an 8-bit string may be used instead of an integer sequence.

In [1]:
from PIL import Image
# take the r, g, b components of the given color
# tint the entire range of colors from 0 to 255
# return a list of ints we'll use later as palette to tint our image
# we use extend instead of append to avoid creating a list of tuples
def make_palette(color):
    palette = []
    r, g, b = color
    for i in range(255):
        red = r * i // 255
        green = g * i // 255
        blue = b * i // 255
        palette.extend((red, green, blue))

    return palette

Help on built-in function extend:

extend(iterable, /) method of builtins.list instance
    Extend list by appending elements from the iterable.



In [10]:
# our tinting reference
# off_white = (2, 240, 192)
off_white = (206, 228, 240)
sepia = make_palette(off_white)
color_image = Image.open('cat.jpg')
gray = color_image.convert("L")
# add sepia tone
gray.putpalette(sepia)
# convert to RGB
rgb_image = gray.convert("RGB")
rgb_image.show()
color_image.show()


In [11]:
# print(make_palette(off_white))
len(make_palette(off_white)) / 3
type(color_image)

PIL.JpegImagePlugin.JpegImageFile